### PR TITLE
Iox #542 resource leak fix

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [x] All checkboxes in the PR checklist are checked or crossed out
1. [x] Merge

## References

fixes partially #542

### CODE REVIEW

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+{
+    bool hasCalledCleanup = false;
+
+    {
+        GenericRAII sut([&] { hasCalledCleanup = true; });
+        GenericRAII sut2(std::move(sut));
+    }
+
+    EXPECT_TRUE(hasCalledCleanup);
+}
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:105]: How do you know that it was called only once?

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        GenericRAII sut([&] { hasCalledCleanup = true; });
+        GenericRAII sut2([&] { hasCalledCleanup2 = true; });
+
+        sut = std::move(sut2);
+    }
+
+    EXPECT_TRUE(hasCalledCleanup);
+    EXPECT_TRUE(hasCalledCleanup2);
+}
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:134]: same here, how do you know that the sut2 dtor doesn't set the cleanup flag

File: iceoryx_utils/source/posix_wrapper/shared_memory_object/memory_map.cpp [hunk 3]
```diff
@@ -123,20 +127,27 @@ MemoryMap& MemoryMap::operator=(MemoryMap&& rhs)
...
+void* MemoryMap::getBaseAddress() const noexcept
 {
     return m_baseAddress;
 }
+
+void MemoryMap::destroy() noexcept
+{
+    if (m_isInitialized)
+    {
+        cxx::makeSmartC(munmap, cxx::ReturnMode::PRE_DEFINED_ERROR_CODE, {-1}, {}, m_baseAddress, m_length);
```
Comment (mossmaurice) [inline iceoryx_utils/source/posix_wrapper/shared_memory_object/memory_map.cpp:147]: What happens if this call fails? Maybe return the error here and inform the user with `std:err` in the d'tor?

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp]: it is a now a counter

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp (hunk 2 above)
Comment (elfenpiff) [reply iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp]: counter

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        GenericRAII sut2(std::move(sut));
+    }
+
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+}
+
+TEST_F(GenericRAII_test, MoveAssignmentCallsCleanup)
+{
+    int hasCalledCleanup = 0;
+    int hasCalledCleanup2 = 0;
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:110]: please either check this or remove it entirely

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+{
+    int hasCalledCleanup = 0;
+
+    {
+        GenericRAII sut([&] { ++hasCalledCleanup; });
+        GenericRAII sut2(std::move(sut));
+    }
+
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+}
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:105]: how do you know that `sut2` calls the cleanup function and not `sut`?

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        GenericRAII sut([&] { ++hasCalledCleanup; });
+        GenericRAII sut2([&] { ++hasCalledCleanup2; });
+
+        sut = std::move(sut2);
+    }
+
+    EXPECT_THAT(hasCalledCleanup, Eq(1));
+    EXPECT_THAT(hasCalledCleanup2, Eq(1));
+}
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:134]: same here ;)

Comment (elfenpiff) [reply iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:134]: Verified by adding an EXPECT_THAT inside the brackets right after move.

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp (hunk 5 above)
Comment (elfenpiff) [reply iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:105]: I added an EXPECT_THAT right after I resetted the sut which I packed into an optional.

**Review by dkroenke — APPROVED** [review]

File: iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        GenericRAII sut([&] { ++hasCalledCleanup; });
+        GenericRAII sut2([&] { ++hasCalledCleanup2; });
+
+        sut = std::move(sut2);
+        EXPECT_THAT(hasCalledCleanup, Eq(1));
+        EXPECT_THAT(hasCalledCleanup2, Eq(0));
+    }
+
+    EXPECT_THAT(hasCalledCleanup2, Eq(1));
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_cxx_generic_raii.cpp:138]: I'm wondering if `hasCalledCleanup` should also be checked here. Technically it's done with the move ctor due to code reuse, but that's just an implementation detail.

**Review by elBoberido — APPROVED** [review]

**Pull request merged**